File: loafer-api/src/main/java/com.roysmond.loafer.api/domain/Post.java
package com.roysmond.loafer.api.domain;

import com.roysmond.loafer.api.domain.emun.PostStatus;
import com.roysmond.loafer.api.domain.emun.PostType;
import lombok.Data;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

/**
 * A Post Or A Issue
 * <p>
 * Created by wuyuexin on 2017/6/3.
 */
@Data
@Entity
@Table(name = "post")
public class Post extends AbstractAuditingModel implements Serializable {

    @NotNull
    @Size(max = 100)
    @Column(name = "title", length = 100, nullable = false)
    private String title;

    @Column(name = "author")
    private String author;

    @Column(name = "summary")
    private String summary;

    @Size(max = 1048576)
    @Column(name = "content", length = 1048576)
    private String content;

    @Enumerated(EnumType.STRING)
    @Column(name = "post_type")
    private PostType type;

    @Enumerated(EnumType.STRING)
    @Column(name = "post_status")
    private PostStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @Column(name = "share_at")
    private LocalDateTime shareAt;

    @Column(name = "vote_up", nullable = false)
    private int voteUp = 0;

    @Column(name = "vote_down", nullable = false)
    private int voteDown = 0;

    @Column(name = "hits", nullable = false)
    private int hits = 0;

    @Column(name = "replies", nullable = false)
    private int replies = 0;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "post_focus",
            joinColumns = {@JoinColumn(name = "post_id", referencedColumnName = "id")},
            inverseJoinColumns = {@JoinColumn(name = "user_id", referencedColumnName = "id")})
    private Set<User> focus = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY, cascade = CascadeType.MERGE)
    @JoinTable(name = "posts_tags",
            joinColumns = {@JoinColumn(name = "post_id", nullable = false, updatable = false)},
            inverseJoinColumns = {@JoinColumn(name = "tag_id", nullable = false, updatable = false)})
    private Set<Tag> tags = new HashSet<>();

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, mappedBy = "post")
    private Set<Vote> votes = new HashSet<>();

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, mappedBy = "post")
    private Set<Reply> replySet = new HashSet<>();

    @OneToMany(orphanRemoval = true, mappedBy = "post")
    private Set<Message> messages = new HashSet<>();

}

File: loafer-webapp/src/main/webapp/src/app/models/Post.ts
import {Tag} from "./Tag";

export class Post {
  private _id: number;
  private _author: string;
  private _title: string;
  private _summary: string;
  private _content: string;
  // https://stackoverflow.com/questions/12870291/typescript-typed-array-usage
  private _tags: Tag[];


  get id(): number {
    return this._id;
  }

  set id(value: number) {
    this._id = value;
  }

  get author(): string {
    return this._author;
  }

  set author(value: string) {
    this._author = value;
  }

  get title(): string {
    return this._title;
  }

  set title(value: string) {
    this._title = value;
  }

  get summary(): string {
    return this._summary;
  }

  set summary(value: string) {
    this._summary = value;
  }

  get content(): string {
    return this._content;
  }

  set content(value: string) {
    this._content = value;
  }

  get tags(): Tag[] {
    return this._tags;
  }

  set tags(value: Tag[]) {
    this._tags = value;
  }
}
